<!doctype html>
<html lang="en">
  <head>
    <meta charset="UTF-8" />
    <title>Tests</title>
    <style>
      html,
      body {
        background-color: black;
      }
    </style>
  </head>
  <body>
    <script type="module">
      import { dumpContext } from "./js/helper.js";
      import {
        testDiskBouncingAgainstWall,
        testDiskBouncingAgainstWallHorizontal,
        testDiskHitArenaCorner,
        testRoundingErrorIsManageable,
        testTouchingDisksOfDifferentVelocityCanSeparate,
        testTwoArenaCollisionInTwoFrames,
      } from "./js/tests.js";
      import update from "./js/update.js";

      {
        const [CONFIG, init, asserts] = testDiskBouncingAgainstWall();
        const CONTEXT = init();

        console.log(dumpContext(CONTEXT));
        for (let i = 0; i < 5; i++) {
          update(CONTEXT, ...CONFIG);
          CONTEXT.elapsed += 1;
          console.log(`====== at the end of ${CONTEXT.elapsed} iteration`);
          console.log(dumpContext(CONTEXT));
          asserts.forEach((a) => a(CONTEXT));
        }
      }

      {
        const [CONFIG, init, asserts] = testDiskBouncingAgainstWallHorizontal();
        const CONTEXT = init();

        console.log(dumpContext(CONTEXT));
        for (let i = 0; i < 5; i++) {
          update(CONTEXT, ...CONFIG);
          CONTEXT.elapsed += 1;
          console.log(`====== at the end of ${CONTEXT.elapsed} iteration`);
          console.log(dumpContext(CONTEXT));
          asserts.forEach((a) => a(CONTEXT));
        }
      }

      {
        const [CONFIG, init, asserts] =
          testTouchingDisksOfDifferentVelocityCanSeparate();
        const CONTEXT = init();

        for (let i = 0; i < 5; i++) {
          update(CONTEXT, ...CONFIG);
          CONTEXT.elapsed += 1;
          const failed = asserts
            .map((a) => a(CONTEXT))
            .some((r) => !!r === false);
          if (failed) {
            console.error("test halted after failed assertion");
            break;
          }
        }
      }

      {
        const [CONFIG, init, asserts] = testTwoArenaCollisionInTwoFrames();
        const CONTEXT = init();

        for (let i = 0; i < 5; i++) {
          update(CONTEXT, ...CONFIG);
          CONTEXT.elapsed += 1;
          const failed = asserts
            .map((a) => a(CONTEXT))
            .some((r) => !!r === false);
          if (failed) {
            console.error("test halted after failed assertion");
            break;
          }
        }
      }

      {
        const [CONFIG, init, asserts] = testRoundingErrorIsManageable();
        const CONTEXT = init();

        for (let i = 0; i < 100; i++) {
          update(CONTEXT, ...CONFIG);
          CONTEXT.elapsed += 1;
          const failed = asserts
            .map((a) => a(CONTEXT))
            .some((r) => !!r === false);
          if (failed) {
            console.error("test halted after failed assertion");
            break;
          }
        }
      }

      {
        const [CONFIG, init, asserts] = testDiskHitArenaCorner();
        const CONTEXT = init();

        for (let i = 0; i < 4; i++) {
          update(CONTEXT, ...CONFIG);
          CONTEXT.elapsed += 1;
          const failed = asserts
            .map((a) => a(CONTEXT))
            .some((r) => !!r === false);
          if (failed) {
            console.error("test halted after failed assertion");
            break;
          }
        }
      }
    </script>
  </body>
</html>
